# HG changeset patch
# User Betterbird <betterbird@betterbird.eu>
# Date 1686684593 -7200
# Parent  6de25d1dd75b064990ca3a4a5f3e49468aa29992
Feature: Rework MailNotificationService.jsm to improve Windows systray tooltip.
* * *
Fix logic in MailNotificationManager.jsm: Not showing badge led to not showing tooltip :-(

diff --git a/mailnews/base/public/mozINewMailListener.idl b/mailnews/base/public/mozINewMailListener.idl
--- a/mailnews/base/public/mozINewMailListener.idl
+++ b/mailnews/base/public/mozINewMailListener.idl
@@ -11,12 +11,13 @@
  * from mozINewMailNotificationService
  * NOTE: THIS INTERFACE IS UNDER ACTIVE DEVELOPMENT AND SUBJECT TO CHANGE,
  * see https://bugzilla.mozilla.org/show_bug.cgi?id=715799
  */
 interface mozINewMailListener : nsISupports {
   /** The new mail notification service will call this when the number of interesting
    * messages has changed
    *
-   * @param unreadCount   The number of unread messages the user cares to be notified about
+   * @param count   The number of unread messages the user cares to be notified about
+   * @param extra   Any additional information that could be relevant (like where the new message are)
    */
-  void onCountChanged(in unsigned long count);
+  void onCountChanged(in unsigned long count, in AString extra);
 };
diff --git a/mailnews/base/src/MailNotificationManager.jsm b/mailnews/base/src/MailNotificationManager.jsm
--- a/mailnews/base/src/MailNotificationManager.jsm
+++ b/mailnews/base/src/MailNotificationManager.jsm
@@ -37,16 +37,17 @@ class MailNotificationManager {
     "nsIFolderListener",
     "mozINewMailListener",
   ]);
 
   constructor() {
     this._systemAlertAvailable = true;
     this._unreadChatCount = 0;
     this._unreadMailCount = 0;
+    this._unreadMailExtra = "";
     // @type {Map<nsIMsgFolder, number>} - A map of folder and its last biff time.
     this._folderBiffTime = new Map();
     // @type {Set<nsIMsgFolder>} - A set of folders to show alert for.
     this._pendingFolders = new Set();
 
     this._logger = console.createInstance({
       prefix: "mail.notification",
       maxLogLevel: "Warn",
@@ -172,19 +173,20 @@ class MailNotificationManager {
   onFolderBoolPropertyChanged() {}
   onFolderUnicharPropertyChanged() {}
   onFolderPropertyFlagChanged() {}
   onFolderEvent() {}
 
   /**
    * @see mozINewMailNotificationService
    */
-  onCountChanged(count) {
+  onCountChanged(count, extra) {
     this._logger.log(`Unread mail count changed to ${count}`);
     this._unreadMailCount = count;
+    this._unreadMailExtra = extra;
     this._updateUnreadCount();
   }
 
   /**
    * Show an alert according to the changed folder.
    *
    * @param {nsIMsgFolder} changedFolder - The folder that emitted the change
    *   event, can be a root folder or a real folder.
@@ -443,25 +445,29 @@ class MailNotificationManager {
     this._updatingUnreadCount = true;
 
     this._logger.debug(
       `Update unreadMailCount=${this._unreadMailCount}, unreadChatCount=${this._unreadChatCount}`
     );
     let count = this._unreadMailCount + this._unreadChatCount;
     let tooltip = "";
     if (AppConstants.platform == "win") {
-      if (!Services.prefs.getBoolPref("mail.biff.show_badge", true)) {
-        count = 0;
-      }
       if (count > 0) {
         tooltip = await lazy.l10n.formatValue("unread-messages-os-tooltip", {
           count,
         });
+        if (this._unreadMailExtra) {
+          tooltip += this._unreadMailExtra;
+        }
       }
-      await lazy.WinUnreadBadge.updateUnreadCount(count, tooltip);
+      if (Services.prefs.getBoolPref("mail.biff.show_badge", true)) {
+        await lazy.WinUnreadBadge.updateUnreadCount(count, tooltip);
+      } else {
+        await lazy.WinUnreadBadge.updateUnreadCount(0, "");
+      }
     }
     this._osIntegration?.updateUnreadCount(count, tooltip);
 
     this._updatingUnreadCount = false;
     if (this._pendingUpdate) {
       // There was at least one _updateUnreadCount call while we were rendering
       // the badge. Render one more time will ensure the badge reflects the
       // current state.
diff --git a/mailnews/base/src/MailNotificationService.jsm b/mailnews/base/src/MailNotificationService.jsm
--- a/mailnews/base/src/MailNotificationService.jsm
+++ b/mailnews/base/src/MailNotificationService.jsm
@@ -21,27 +21,30 @@ var { MailServices } = ChromeUtils.impor
  * @implements {nsIFolderListener}
  * @implements {nsIObserver}
  */
 function NewMailNotificationService() {
   this._mUnreadCount = 0;
   this._mNewCount = 0;
   this._listeners = [];
   this.wrappedJSObject = this;
+  this.folderCountMap = new Map();
+  this.folderNewMap = new Map();
 
   this._log = console.createInstance({
     prefix: "mail.notification",
     maxLogLevel: "Warn",
     maxLogLevelPref: "mail.notification.loglevel",
   });
 
   Services.obs.addObserver(this, "profile-before-change");
   MailServices.mailSession.AddFolderListener(
     this,
     Ci.nsIFolderListener.intPropertyChanged |
+      Ci.nsIFolderListener.boolPropertyChanged |
       Ci.nsIFolderListener.added |
       Ci.nsIFolderListener.removed |
       Ci.nsIFolderListener.propertyFlagChanged
   );
   this._initUnreadCount();
 }
 
 NewMailNotificationService.prototype = {
@@ -134,16 +137,21 @@ NewMailNotificationService.prototype = {
     } else if (
       aFolder.flags & Ci.nsMsgFolderFlags.SpecialUse &&
       !(aFolder.flags & Ci.nsMsgFolderFlags.Inbox)
     ) {
       // It's a special folder *other than the inbox*, don't count it by default.
       shouldCount.data = false;
     } else if (aFolder.flags & Ci.nsMsgFolderFlags.Virtual) {
       shouldCount.data = false;
+    } else if (
+      Services.prefs.getBoolPref("mail.biff.use_new_count_in_badge", true)
+    ) {
+      // If we're counting new messages, we count them in all folders.
+      shouldCount.data = true;
     } else {
       // If we're only counting inboxes and it's not an inbox...
       try {
         // If we can't get this pref, just leave it as the default
         let onlyCountInboxes = Services.prefs.getBoolPref(
           "mail.notification.count.inbox_only"
         );
         if (onlyCountInboxes && !(aFolder.flags & Ci.nsMsgFolderFlags.Inbox)) {
@@ -156,104 +164,216 @@ NewMailNotificationService.prototype = {
     Services.obs.notifyObservers(
       shouldCount,
       "before-count-unread-for-folder",
       aFolder.URI
     );
     return shouldCount.data;
   },
 
+  onFolderBoolPropertyChanged(folder, property, oldValue, newValue) {
+    this._log.trace(
+      `Changed bool ${property} of ${folder.folderURL}: ${oldValue} -> ${newValue}`
+    );
+    if (property == "NewMessages") {
+      this._newMessagesChanged(folder, oldValue, newValue);
+    }
+  },
+
   onFolderIntPropertyChanged(folder, property, oldValue, newValue) {
     try {
       if (property == "FolderSize") {
         return;
       }
       this._log.trace(
         `Changed int ${property} of ${folder.folderURL}: ${oldValue} -> ${newValue}`
       );
-      if (property == "BiffState") {
-        this._biffStateChanged(folder, oldValue, newValue);
-      } else if (property == "TotalUnreadMessages") {
+      if (property == "TotalUnreadMessages") {
         this._updateUnreadCount(folder, oldValue, newValue);
-      } else if (property == "NewMailReceived") {
-        this._newMailReceived(folder, oldValue, newValue);
+        // For new folders, there is a call with old == -1 and new == 0.
+        this._updateNewCount(folder, newValue - Math.max(oldValue, 0));
       }
     } catch (error) {
       this._log.error("onFolderIntPropertyChanged: exception " + error);
     }
   },
 
-  _biffStateChanged(folder, oldValue, newValue) {
-    if (newValue == Ci.nsIMsgFolder.nsMsgBiffState_NewMail) {
-      if (folder.server && !folder.server.performingBiff) {
-        this._log.debug(
-          `${folder.URI} notified, but server not performing biff`
+  _updateNewCount(folder, newCount, allowNegative = false) {
+    if (!this.confirmShouldCount(folder)) {
+      return;
+    }
+
+    this._log.debug(`_updateNewCount: ${folder.URI} new messages: ${newCount}`);
+    let folderCount = this.folderCountMap.get(folder.URI);
+    if (!folderCount) {
+      if (allowNegative) {
+        this.folderCountMap.set(folder.URI, newCount);
+      } else {
+        this.folderCountMap.set(folder.URI, Math.max(newCount, 0));
+      }
+    } else {
+      /* eslint-disable no-lonely-if */
+      if (allowNegative) {
+        this.folderCountMap.set(folder.URI, folderCount + newCount);
+      } else {
+        this.folderCountMap.set(
+          folder.URI,
+          Math.max(folderCount + newCount, 0)
         );
-        return;
       }
+      /* eslint-enable no-lonely-if */
+    }
 
-      // Biff notifications come in for the top level of the server, we need to
-      // look for the folder that actually contains the new mail.
+    let folderNew = this.folderNewMap.get(folder.URI);
+    if (!folderNew) {
+      this.folderNewMap.set(folder.URI, {
+        name: folder.name,
+        isInbox: folder.flags & Ci.nsMsgFolderFlags.Inbox,
+      });
+    }
 
-      let allFolders = [folder, ...folder.descendants];
-
-      this._log.debug(`${folder.URI} notified; will check subfolders`);
-      let newCount = 0;
+    this._mNewCount = Math.max(this._mNewCount + newCount, 0);
+    this._log.debug(
+      `_updateNewCount: ${folder.URI} new mail count: ${this._mNewCount}`
+    );
+    if (this.countNew) {
+      this._notifyListeners(
+        Ci.mozINewMailNotificationService.count,
+        "onCountChanged",
+        this._mNewCount,
+        this._folderInfo()
+      );
+    }
+  },
 
-      for (let folder of allFolders) {
-        if (this.confirmShouldCount(folder)) {
-          let folderNew = folder.getNumNewMessages(false);
-          this._log.debug(`${folder.URI}: ${folderNew} new`);
-          if (folderNew > 0) {
-            newCount += folderNew;
+  _newMessagesChanged(folder, oldValue, newValue) {
+    if (newValue) {
+      // We have new messages.
+      // `getNumNewMessages()` doesn't work until after the biff
+      // notification (BiffState or NumNewBiffMessages changed)
+      // and it doesn't work on local folders at all.
+      // We use a change of TotalUnreadMessages.
+      let folderNew = this.folderNewMap.get(folder.URI);
+      if (!folderNew) {
+        this.folderNewMap.set(folder.URI, {
+          name: folder.name,
+          isInbox: folder.flags & Ci.nsMsgFolderFlags.Inbox,
+        });
+        let folderCount = this.folderCountMap.get(folder.URI);
+        if (folderCount) {
+          // The count arrived before the "NewMessages" notification.
+          this._mNewCount += folderCount;
+          this._log.debug(
+            `_newMessagesChanged: ${folder.URI} new mail count: ${this._mNewCount}`
+          );
+          if (this.countNew) {
+            this._notifyListeners(
+              Ci.mozINewMailNotificationService.count,
+              "onCountChanged",
+              this._mNewCount,
+              this._folderInfo()
+            );
           }
         }
       }
-      if (newCount > 0) {
-        this._mNewCount += newCount;
-        this._log.debug(`${folder.URI}: new mail count ${this._mNewCount}`);
+    } else {
+      // New messages went away.
+      let folderCount = this.folderCountMap.get(folder.URI);
+      this.folderCountMap.delete(folder.URI);
+      this.folderNewMap.delete(folder.URI);
+      if (folderCount) {
+        this._mNewCount = Math.max(this._mNewCount - folderCount, 0);
+        this._log.debug(
+          `_newMessagesChanged: ${folder.URI} new mail count: ${this._mNewCount}`
+        );
         if (this.countNew) {
           this._notifyListeners(
             Ci.mozINewMailNotificationService.count,
             "onCountChanged",
-            this._mNewCount
+            this._mNewCount,
+            this._folderInfo()
           );
         }
       }
-    } else if (newValue == Ci.nsIMsgFolder.nsMsgBiffState_NoMail) {
-      // Dodgy - when any folder tells us it has no mail, clear all unread mail
-      this._mNewCount = 0;
-      this._log.debug(`${folder.URI}: no new mail`);
-      if (this.countNew) {
-        this._notifyListeners(
-          Ci.mozINewMailNotificationService.count,
-          "onCountChanged",
-          this._mNewCount
-        );
-      }
     }
   },
 
-  _newMailReceived(folder, oldValue, newValue) {
-    if (!this.confirmShouldCount(folder)) {
-      return;
+  _folderInfo() {
+    let totalInbox = 0;
+    let countInbox = 0;
+    let inboxName = "";
+    let folders = [];
+    for (const [key, value] of this.folderNewMap) {
+      if (!this.folderNewMap.get(key)) {
+        continue;
+      }
+      let count = this.folderCountMap.get(key);
+      if (!count) {
+        continue;
+      }
+      if (value.isInbox) {
+        totalInbox += count;
+        countInbox++;
+        inboxName = value.name; // Could be localised.
+        continue;
+      }
+      folders.push({ count, name: value.name });
     }
-
-    if (!oldValue || oldValue < 0) {
-      oldValue = 0;
+    if (totalInbox == 0 && folders.length == 0) {
+      return "";
+    }
+    let output = "\n";
+    let folderAddsComma = false;
+    if (totalInbox > 0) {
+      output += inboxName;
+      if (countInbox > 1) {
+        output += "(Σ)";
+      }
+      output += ": " + totalInbox.toString(10);
+      folderAddsComma = true;
+    }
+    if (folders.length == 0) {
+      return output;
     }
-    this._mNewCount += newValue - oldValue;
-    this._log.debug(`_newMailReceived ${folder.URI} - ${this._mNewCount} new`);
-    if (this.countNew) {
-      this._notifyListeners(
-        Ci.mozINewMailNotificationService.count,
-        "onCountChanged",
-        this._mNewCount
-      );
+    let lineLength = output.length - 1; // Don't count the initial newline.
+    folders.sort((el1, el2) => el2.count - el1.count);
+    for (let i = 0; i < folders.length; i++) {
+      let folder = folders[i].name + ": " + folders[i].count.toString(10);
+      // The total length of the tooltip must no exceed 127 characters, or it's truncated.
+      // There are already around 27 characters in "9999 unread messages" or
+      // "9999 ungelesene Nachrichten", other languages may have more
+      // (although I didn't see one, checked es, pt, zh, ja).
+      // So let's keep the count unter 100. We need two characters for
+      // the connector, either ", " or "\n ".
+      if (output.length + folder.length + 2 > 100) {
+        // Sorry, can't list more folders.
+        output += ", \u2026";
+        break;
+      }
+      if (lineLength + folder.length > 50) {
+        if (folderAddsComma) {
+          output += ",\n" + folder;
+        } else {
+          // No comma, so this is the first thing we add.
+          output += folder;
+          folderAddsComma = true;
+        }
+        lineLength = folder.length;
+        continue;
+      }
+      if (folderAddsComma) {
+        output += ", " + folder;
+      } else {
+        // No comma, so this is the first thing we add.
+        output += folder;
+        folderAddsComma = true;
+      }
+      lineLength += folder.length;
     }
+    return output;
   },
 
   _updateUnreadCount(folder, oldValue, newValue) {
     if (!this.confirmShouldCount(folder)) {
       return;
     }
 
     // treat "count unknown" as zero
@@ -286,28 +406,52 @@ NewMailNotificationService.prototype = {
 
   onMessageAdded(parentFolder, msg) {
     if (this.confirmShouldCount(msg.folder)) {
       this._log.trace(`Added <${msg.messageId}> to ${msg.folder.folderURL}`);
     }
   },
 
   onFolderPropertyFlagChanged(msg, property, oldFlag, newFlag) {
+    if (property != "Status") {
+      return;
+    }
+    this._log.trace(
+      `Changed flag ${property} of ${msg.folder.folderURL}: ${oldFlag} -> ${newFlag}`
+    );
     if (
       oldFlag & Ci.nsMsgMessageFlags.New &&
       !(newFlag & Ci.nsMsgMessageFlags.New)
     ) {
       this._log.trace(
         `<${msg.messageId}> marked read in ${msg.folder.folderURL}`
       );
     } else if (newFlag & Ci.nsMsgMessageFlags.New) {
       this._log.trace(
         `<${msg.messageId}> marked unread in ${msg.folder.folderURL}`
       );
     }
+
+    if (!(oldFlag & Ci.nsMsgMessageFlags.New)) {
+      // Changing read status of non-new message. We need to
+      // compensate since this will change the (new) unread
+      // message count which we might be tracking on the folder.
+      if (
+        oldFlag & Ci.nsMsgMessageFlags.Read &&
+        !(newFlag & Ci.nsMsgMessageFlags.Read)
+      ) {
+        this._updateNewCount(msg.folder, -1, true);
+      }
+      if (
+        !(oldFlag & Ci.nsMsgMessageFlags.Read) &&
+        newFlag & Ci.nsMsgMessageFlags.Read
+      ) {
+        this._updateNewCount(msg.folder, 1);
+      }
+    }
   },
 
   onFolderRemoved(parentFolder, child) {
     if (child.rootFolder == child) {
       this._log.trace(`Removed root folder ${child.folderURL}`);
     } else {
       this._log.trace(
         `Removed child folder ${child.folderURL} from ${parentFolder?.folderURL}`
@@ -338,18 +482,21 @@ NewMailNotificationService.prototype = {
       if (l.obj === aListener) {
         l.flags = flags;
         return;
       }
     }
 
     // Ensure that first-time listeners get an accurate mail count.
     if (flags & Ci.mozINewMailNotificationService.count) {
-      const count = this.countNew ? this._mNewCount : this._mUnreadCount;
-      aListener.onCountChanged(count);
+      if (this.countNew) {
+        aListener.onCountChanged(this._mNewCount, this._folderInfo());
+      } else {
+        aListener.onCountChanged(this._mUnreadCount);
+      }
     }
 
     // If we get here, the listener wasn't already in the list
     this._listeners.push({ obj: aListener, flags });
   },
 
   removeListener(aListener) {
     for (let i = 0; i < this._listeners.length; i++) {
@@ -367,15 +514,15 @@ NewMailNotificationService.prototype = {
       let l = this._listeners[i];
       if (l.flags & flag) {
         list.push(l.obj);
       }
     }
     return list;
   },
 
-  _notifyListeners(flag, func, value) {
+  _notifyListeners(flag, func, value1, value2 = "") {
     let list = this._listenersForFlag(flag);
     for (let i = 0; i < list.length; i++) {
-      list[i][func](value);
+      list[i][func](value1, value2);
     }
   },
 };
diff --git a/mailnews/base/src/WinUnreadBadge.jsm b/mailnews/base/src/WinUnreadBadge.jsm
--- a/mailnews/base/src/WinUnreadBadge.jsm
+++ b/mailnews/base/src/WinUnreadBadge.jsm
@@ -75,17 +75,17 @@ function drawUnreadCountText(cxt, text) 
   cxt.save();
 
   let imageSize = cxt.canvas.width;
 
   // Use smaller fonts for longer text to try and squeeze it in.
   let fontSize = imageSize * (0.95 - 0.15 * text.length);
 
   cxt.font = "500 " + fontSize + "px Calibri";
-  cxt.fillStyle = "#ffffff";
+  cxt.fillStyle = "#000000";
   cxt.textAlign = "center";
 
   // TODO: There isn't a textBaseline for accurate vertical centering ('middle' is the
   // middle of the 'em block', and digits extend higher than 'm'), and the Mozilla core
   // does not currently support computation of ascenders and descenters in measureText().
   // So, we just assume that the font is 70% of the 'px' height we requested, then
   // compute where the baseline ought to be located.
   let approximateHeight = fontSize * 0.7;
@@ -108,17 +108,17 @@ function drawUnreadCountText(cxt, text) 
  */
 function createModernBadgeStyle(canvas, text) {
   let cxt = canvas.getContext("2d");
   let iconSize = canvas.width;
 
   // Draw the background.
   cxt.save();
   // Solid color first.
-  cxt.fillStyle = "#ff0039";
+  cxt.fillStyle = "#ffffff";
   cxt.shadowOffsetX = 0;
   cxt.shadowOffsetY = 0;
   cxt.shadowColor = "rgba(0,0,0,0.7)";
   cxt.shadowBlur = iconSize / 10;
   cxt.beginPath();
   cxt.arc(iconSize / 2, iconSize / 2, iconSize / 2.25, 0, Math.PI * 2, true);
   cxt.fill();
   cxt.clip();
diff --git a/mailnews/base/src/nsMessengerWinIntegration.cpp b/mailnews/base/src/nsMessengerWinIntegration.cpp
--- a/mailnews/base/src/nsMessengerWinIntegration.cpp
+++ b/mailnews/base/src/nsMessengerWinIntegration.cpp
@@ -262,20 +262,21 @@ nsresult nsMessengerWinIntegration::SetT
     NS_ENSURE_TRUE(bundleService, NS_ERROR_UNEXPECTED);
     nsCOMPtr<nsIStringBundle> bundle;
     rv = bundleService->CreateBundle(
         "chrome://branding/locale/brand.properties", getter_AddRefs(bundle));
     NS_ENSURE_SUCCESS(rv, rv);
     rv = bundle->GetStringFromName("brandShortName", mBrandShortName);
     NS_ENSURE_SUCCESS(rv, rv);
   }
-  nsString tooltip = mBrandShortName;
-  if (!mUnreadTooltip.IsEmpty()) {
-    tooltip.AppendLiteral("\n");
-    tooltip.Append(mUnreadTooltip);
+  nsString tooltip;
+  if (mUnreadTooltip.IsEmpty()) {
+    tooltip = mBrandShortName;
+  } else {
+    tooltip = mUnreadTooltip;
   }
   size_t destLength =
       sizeof sMailIconData.szTip / (sizeof sMailIconData.szTip[0]);
   ::StringCchCopyNW(sMailIconData.szTip, destLength, tooltip.get(),
                     tooltip.Length());
   return rv;
 }
 
